File: sdk/typescript/src/wallet/resources/wallets.ts
/**
 * inkbox-wallet/resources/wallets.ts
 *
 * Wallet operations: create, list, balance, send, sign auth, history, and pay-request.
 */

import { HttpTransport } from "../../_http.js";
import {
  AgentWallet,
  AgentWalletBalance,
  OnchainTransactionPage,
  RawAgentWallet,
  RawAgentWalletBalance,
  RawOnchainTransactionPage,
  RawWalletAuthSignature,
  RawWalletPayRequestResponse,
  RawWalletTransaction,
  RawWalletTransactionReceipt,
  WalletAuthSignature,
  WalletPayRequestResponse,
  WalletTransaction,
  WalletTransactionReceipt,
  parseAgentWallet,
  parseAgentWalletBalance,
  parseOnchainTransactionPage,
  parseWalletAuthSignature,
  parseWalletPayRequestResponse,
  parseWalletTransaction,
  parseWalletTransactionReceipt,
} from "../types.js";

export class WalletsResource {
  constructor(private readonly http: HttpTransport) {}

  /** List wallets visible to the caller. */
  async list(): Promise<AgentWallet[]> {
    const data = await this.http.get<RawAgentWallet[]>("/");
    return data.map(parseAgentWallet);
  }

  /** Create a new wallet for an identity. */
  async create(options: {
    agentHandle: string;
    chains?: string[];
  }): Promise<AgentWallet> {
    const body: Record<string, unknown> = {
      agent_handle: options.agentHandle,
    };
    if (options.chains !== undefined) body["chains"] = options.chains;
    const data = await this.http.post<RawAgentWallet>("/", body);
    return parseAgentWallet(data);
  }

  /** Fetch a wallet by ID. */
  async get(walletId: string): Promise<AgentWallet> {
    const data = await this.http.get<RawAgentWallet>(`/${walletId}`);
    return parseAgentWallet(data);
  }

  /** Fetch live on-chain balances for a wallet. */
  async getBalance(walletId: string): Promise<AgentWalletBalance> {
    const data = await this.http.get<RawAgentWalletBalance>(`/${walletId}/balance`);
    return parseAgentWalletBalance(data);
  }

  /** Broadcast an outbound transaction from a wallet. */
  async send(
    walletId: string,
    options: {
      chain: string;
      toAddress: string;
      token: string;
      amount: string;
      memo?: string;
      idempotencyKey?: string;
    },
  ): Promise<WalletTransaction> {
    const body: Record<string, unknown> = {
      chain: options.chain,
      to_address: options.toAddress,
      token: options.token,
      amount: options.amount,
    };
    if (options.memo !== undefined) body["memo"] = options.memo;
    if (options.idempotencyKey !== undefined) body["idempotency_key"] = options.idempotencyKey;
    const data = await this.http.post<RawWalletTransaction>(`/${walletId}/send`, body);
    return parseWalletTransaction(data);
  }

  /** Sign a SIWE-style authentication challenge. */
  async signAuth(
    walletId: string,
    options: { message: string },
  ): Promise<WalletAuthSignature> {
    const data = await this.http.post<RawWalletAuthSignature>(
      `/${walletId}/sign-auth`,
      { message: options.message },
    );
    return parseWalletAuthSignature(data);
  }

  /** List wallet transactions from the server-side audit log. */
  async listTransactions(
    walletId: string,
    options: {
      chain?: string;
      status?: string;
      limit?: number;
    } = {},
  ): Promise<WalletTransaction[]> {
    const data = await this.http.get<RawWalletTransaction[]>(
      `/${walletId}/transactions`,
      {
        chain: options.chain,
        status: options.status,
        limit: options.limit,
      },
    );
    return data.map(parseWalletTransaction);
  }

  /** Fetch the current on-chain receipt state for one transaction row. */
  async getTransactionReceipt(
    walletId: string,
    transactionId: string,
  ): Promise<WalletTransactionReceipt> {
    const data = await this.http.get<RawWalletTransactionReceipt>(
      `/${walletId}/transactions/${transactionId}/receipt`,
    );
    return parseWalletTransactionReceipt(data);
  }

  /** List read-through on-chain history for a wallet. */
  async listOnchainTransactions(
    walletId: string,
    options: {
      chain?: string;
      direction?: string;
      cursor?: string;
      limit?: number;
    } = {},
  ): Promise<OnchainTransactionPage> {
    const data = await this.http.get<RawOnchainTransactionPage>(
      `/${walletId}/onchain-transactions`,
      {
        chain: options.chain,
        direction: options.direction,
        cursor: options.cursor,
        limit: options.limit,
      },
    );
    return parseOnchainTransactionPage(data);
  }

  /** Make an HTTP request and automatically pay any supported 402 challenge. */
  async payRequest(
    walletId: string,
    options: {
      url: string;
      method?: string;
      headers?: Record<string, string>;
      bodyBase64?: string;
      maxCost?: string | number;
    },
  ): Promise<WalletPayRequestResponse> {
    const body: Record<string, unknown> = {
      url: options.url,
    };
    if (options.method !== undefined) body["method"] = options.method;
    if (options.headers !== undefined) body["headers"] = options.headers;
    if (options.bodyBase64 !== undefined) body["body"] = options.bodyBase64;
    if (options.maxCost !== undefined) body["max_cost"] = String(options.maxCost);

    const data = await this.http.post<RawWalletPayRequestResponse>(
      `/${walletId}/pay-request`,
      body,
    );
    return parseWalletPayRequestResponse(data);
  }
}

File: sdk/python/inkbox/wallet/resources/wallets.py
"""
inkbox/wallet/resources/wallets.py

Wallet operations: create, list, balance, send, sign auth, history, and pay-request.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from uuid import UUID

from inkbox.wallet.types import (
    AgentWallet,
    AgentWalletBalance,
    WalletAuthSignature,
    WalletPayRequestResponse,
    WalletTransaction,
    WalletTransactionReceipt,
)

if TYPE_CHECKING:
    from inkbox._http import HttpTransport


class WalletsResource:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def list(self) -> list[AgentWallet]:
        """List wallets visible to the caller."""
        data = self._http.get("/")
        return [AgentWallet._from_dict(item) for item in data]

    def create(
        self,
        *,
        agent_handle: str,
        chains: list[str] | None = None,
    ) -> AgentWallet:
        """Create a new wallet for an identity."""
        body: dict[str, Any] = {"agent_handle": agent_handle}
        if chains is not None:
            body["chains"] = chains
        data = self._http.post("/", json=body)
        return AgentWallet._from_dict(data)

    def get(self, wallet_id: UUID | str) -> AgentWallet:
        """Fetch a wallet by ID."""
        data = self._http.get(f"/{wallet_id}")
        return AgentWallet._from_dict(data)

    def get_balance(self, wallet_id: UUID | str) -> AgentWalletBalance:
        """Fetch live on-chain balances for a wallet."""
        data = self._http.get(f"/{wallet_id}/balance")
        return AgentWalletBalance._from_dict(data)

    def send(
        self,
        wallet_id: UUID | str,
        *,
        chain: str,
        to_address: str,
        token: str,
        amount: str,
        memo: str | None = None,
        idempotency_key: str | None = None,
    ) -> WalletTransaction:
        """Broadcast an outbound transaction from a wallet."""
        body: dict[str, Any] = {
            "chain": chain,
            "to_address": to_address,
            "token": token,
            "amount": amount,
        }
        if memo is not None:
            body["memo"] = memo
        if idempotency_key is not None:
            body["idempotency_key"] = idempotency_key
        data = self._http.post(f"/{wallet_id}/send", json=body)
        return WalletTransaction._from_dict(data)

    def sign_auth(
        self,
        wallet_id: UUID | str,
        *,
        message: str,
    ) -> WalletAuthSignature:
        """Sign a SIWE-style authentication challenge."""
        data = self._http.post(
            f"/{wallet_id}/sign-auth",
            json={"message": message},
        )
        return WalletAuthSignature._from_dict(data)

    def list_transactions(
        self,
        wallet_id: UUID | str,
        *,
        chain: str | None = None,
        status: str | None = None,
        limit: int | None = None,
    ) -> list[WalletTransaction]:
        """List wallet transactions from the server-side audit log."""
        data = self._http.get(
            f"/{wallet_id}/transactions",
            params={"chain": chain, "status": status, "limit": limit},
        )
        return [WalletTransaction._from_dict(item) for item in data]

    def get_transaction_receipt(
        self,
        wallet_id: UUID | str,
        transaction_id: UUID | str,
    ) -> WalletTransactionReceipt:
        """Fetch the current on-chain receipt state for one transaction row."""
        data = self._http.get(f"/{wallet_id}/transactions/{transaction_id}/receipt")
        return WalletTransactionReceipt._from_dict(data)

    def pay_request(
        self,
        wallet_id: UUID | str,
        *,
        url: str,
        method: str | None = None,
        headers: dict[str, str] | None = None,
        body_base64: str | None = None,
        max_cost: str | int | float | None = None,
    ) -> WalletPayRequestResponse:
        """Make an HTTP request and automatically pay any supported 402 challenge."""
        body: dict[str, Any] = {"url": url}
        if method is not None:
            body["method"] = method
        if headers is not None:
            body["headers"] = headers
        if body_base64 is not None:
            body["body"] = body_base64
        if max_cost is not None:
            body["max_cost"] = str(max_cost)
        data = self._http.post(f"/{wallet_id}/pay-request", json=body)
        return WalletPayRequestResponse._from_dict(data)
